### PR TITLE
Added basic block source lines to source_locationt

### PR DESCRIPTION
**Pull request by tautschnig**
This is a rebased version of #1439.

Block coverage obtained with "cbmc --cover locations" now reports the file name
and line number of every line of source code contributing to a basic block in
the "description" field of the xml output. (The lines contributing to a block
can come from multiple files via function inlining and definitions in include
files, so reporting line numbers alone is not sufficient.)

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
Seems reasonable.  I wonder if @peterschrammel or whoever is in charge of Diffblue's test case generator wants to comment as this seems like the kind of thing that would change their output.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 31727ed0563750e31c1112397baa6b4888df477a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98753027

File: src/goto-instrument/source_lines.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <util/string2int.h>
+#include <util/string_utils.h>
+
+#include <sstream>
+
+void source_linest::insert(source_locationt const &loc)
+{
+  if(loc.get_file().empty() || loc.is_built_in())
+    return;
+  std::string file = id2string(loc.get_file());
```
Comment (smowton) [inline src/goto-instrument/source_lines.cpp]: Const ref?

File: src/goto-instrument/source_lines.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  // the function of a source location can fail to be set
+  std::string func = id2string(loc.get_function());
```
Comment (smowton) [inline src/goto-instrument/source_lines.cpp]: const ref?

**Review by smowton — APPROVED** [review]

File: src/goto-instrument/source_lines.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/source_lines.cpp]: Indeed, done.

File: src/goto-instrument/source_lines.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-instrument/source_lines.cpp]: Done.

File: regression/cbmc-cover/block-coverage-report2/test.desc [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+main.c
+--cover location
+block 1 \(lines main.c:main:13,14\): SATISFIED
+block 2 \(lines main.c:main:15\): SATISFIED
+block 3 \(lines main.c:main:17,18\): SATISFIED
+block 4 \(lines main.c:main:18,19\): SATISFIED
+block 5 \(lines main.c:main:20\): SATISFIED
+block 6 \(lines main.c:main:15,21,22\): SATISFIED
+block 7 \(lines main.c:main:24,25\): SATISFIED
+block 1 \(lines main.c:foo:6,7,8,9\): SATISFIED
```
Comment (peterschrammel) [inline regression/cbmc-cover/block-coverage-report2/test.desc]: Suggest to use `util/format_number_range` for a more compact representation

Comment (tautschnig) [reply regression/cbmc-cover/block-coverage-report2/test.desc]: Yes, we should do that. As Mark explained in #1439 we have a tool (to be open sourced) that uses this information, so I'll first make sure it can cope with both the explicit enumeration as well as the ranges.

Comment (tautschnig) [reply regression/cbmc-cover/block-coverage-report2/test.desc]: Done.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: b70f6965ba30a9fdbf4f47584e11c9e66b3e54a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99099532
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: regression/cbmc-cover/block-coverage-report2/test.desc (hunk 3 above)
Comment (markrtuttle) [reply regression/cbmc-cover/block-coverage-report2/test.desc]: In practice, blocks are so small it makes almost no difference.  

But the essential thing is to continue listing filename-numberset pairs, because the lines contributing to a block can come from more than one file.

Comment (tautschnig) [reply regression/cbmc-cover/block-coverage-report2/test.desc]: That should be preserved, yes. The only change (and really the only change that was necessary to the regression tests) is that instead of `6,7,8,9` we now print `6-9`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b3209a0253973b6e67951854e04f53c659b10bda).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99113365

**Review by peterschrammel — APPROVED** [review]

**Comment by tautschnig** [comment]
Marking do-not-merge: I would like to ask for explicit approval that this doesn't break TG (as this PR changes the output) and also I need to co-ordinate the merge on our end.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b1a2804f6a4b1a795155f45db83392db01c93f72).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99575601

**Comment by jeannielynnmoulton** [comment]
Bump is here: https://github.com/diffblue/test-gen/pull/2968. I think this PR will need rebasing on cbmc develop to pass all CI on test-gen (due to https://github.com/diffblue/cbmc/pull/4067)

**Comment by tautschnig** [comment]
Thanks @jeannielynnmoulton! This PR is now rebased.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 77ae56c9b8e7e3f719feac0f2e1127e7308995fc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99673839
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a51d007f5a6c760b8887c11f5824e92561232861).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99704188

**Comment by jeannielynnmoulton** [comment]
@tautschnig I've had to update some tests on test-gen for this change, but it doesn't seem to break anything. The one curious thing that I see is that some `lines` are blank like this:

`[java::Test.privateStatic:(I)I.coverage.1] file Test.java line 28 function java::Test.privateStatic:(I)I bytecode-index 1 block 1 (lines ): SATISFIED`

Is this expected?

**Comment by jeannielynnmoulton** [comment]
Bump passes. It would be good to get an answer to the above question before this is merged though.

**Comment by tautschnig** [comment]
Thanks @jeannielynnmoulton - I'll look into this, it is reproducible on one of the regression tests.

**Comment by tautschnig** [comment]
@jeannielynnmoulton Thank you very much for not letting me be lazy - I have now added code to `cover_basic_blocks_javat` to actually hold the desired info, not just the non-Java one... May I ask you to do another test run as the output should now be non-empty? (It certainly is non-empty in the jdiff test included in the repository.)

**Comment by jeannielynnmoulton** [comment]
@tautschnig Seems to be non-empty now. `[java::Test.privateStatic:(I)I.coverage.1] file Test.java line 28 function java::Test.privateStatic:(I)I bytecode-index 1 block 1 (lines Test.java:java::Test.privateStatic:(I)I:28): SATISFIED`

**Comment by tautschnig** [comment]
@jeannielynnmoulton Many thanks for confirming! Does it still pass TG tests, or should I hold back on merging this one (once Travis has settled)?

**Comment by jeannielynnmoulton** [comment]
I'm updating the tests to be more specific so we don't have the blank lines regression. I'll run CI one more time to be safe and let you know when it's safe to merge. (Sorry about closing the PR accidentally, my laptop track pad is so sensitive)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 01f4ee2672cf97491ff5fe8e17b09457dba61a96).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100217199

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 01f4ee2672cf97491ff5fe8e17b09457dba61a96).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100219604

**Comment by jeannielynnmoulton** [comment]
@tautschnig Bump passed!
